oak_tensorflow_service/src/tflite.rs

```diff
@@ -66,51 +70,55 @@ const TENSOR_ARENA_SIZE: usize = 1024;
 #[derive(Default)]
 pub struct TfliteModel {
     tensor_arena: Vec<u8>,
+    output_buffer_len: Option<usize>,
 }
 
 impl TfliteModel {
     pub fn new() -> Self {
         let tensor_arena = vec![0; TENSOR_ARENA_SIZE];
-        Self { tensor_arena }
+        Self {
+            tensor_arena,
+            output_buffer_len: None,
+        }
     }
 
-    pub fn initialize(&self, model_bytes: &[u8]) -> anyhow::Result<()> {
-        let model_bytes_len = model_bytes
-            .len()
-            .try_into()
-            .map_err(|error| anyhow!("Failed to convert usize to u64: {}", error))?;
-        let tensor_arena_len = self
-            .tensor_arena
-            .len()
-            .try_into()
-            .map_err(|error| anyhow!("Failed to convert usize to u64: {}", error))?;
+    pub fn initialize(&mut self, model_bytes: &[u8]) -> anyhow::Result<()> {
+        let model_bytes_len = model_bytes.len();
+        let tensor_arena_len = self.tensor_arena.len();
+        let mut output_buffer_len = 0usize;
         let _ = unsafe {
             tflite_init(
                 model_bytes.as_ptr(),
                 model_bytes_len,
                 self.tensor_arena.as_ptr(),
                 tensor_arena_len,
+                &mut output_buffer_len,
             )
         };
+        self.output_buffer_len = Some(output_buffer_len);
         Ok(())
     }
 
-    pub fn run(&self, input_bytes: &[u8]) -> anyhow::Result<Vec<u8>> {
-        let input_bytes_len = input_bytes
-            .len()
-            .try_into()
-            .map_err(|error| anyhow!("Failed to convert usize to u64: {}", error))?;
-        // TODO(#3297): Allocate enough bytes for the TensorFlow Lite model.
-        let mut output_bytes = vec![0; input_bytes.len()];
-        let mut output_bytes_len = 0u64;
+    pub fn run(&mut self, input_bytes: &[u8]) -> anyhow::Result<Vec<u8>> {
+        let buffer_len = self
+            .output_buffer_len
+            .context("Running inference on a non-initialized TensorFlow model")?;
+
+        let input_bytes_len = input_bytes.len();
+        let mut output_bytes = vec![0; buffer_len];
+        let mut output_bytes_len = 0usize;
         let _ = unsafe {
             tflite_run(
                 input_bytes.as_ptr(),
                 input_bytes_len,
                 output_bytes.as_mut_ptr(),
-                &mut output_bytes_len as *mut u64,
+                &mut output_bytes_len,
             )
         };
-        Ok(output_bytes)
+        if output_bytes_len <= buffer_len {
+            Ok(output_bytes[..output_bytes_len].to_vec())
+        } else {
+            Err(anyhow!("Response size is larger than the allocated buffer"))
```

Comment:
I would go as far as panicking in this case. If the c++ code wrote a response larger than the buffer it means there is a potential for memory corruption and we can't rely on any of the Rust memory safety assumptions any more.

Reply:
Done